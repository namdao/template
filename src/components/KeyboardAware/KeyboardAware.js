import React from 'react';
import PropTypes from 'prop-types';
import { Animated, Keyboard, ViewPropTypes } from 'react-native';
import Platform from 'utils/platform';
import styles from './styles';

const { KeyboardEvent } = Platform;

export default class KeyboardAwareComponent extends React.Component {
  static propTypes = {
    children: PropTypes.oneOfType(PropTypes.node, PropTypes.func),
    style: ViewPropTypes.style,
  };

  static defaultProps = {
    children: null,
    style: {},
  };

  constructor(props) {
    super(props);
    this.keyboardHeight = new Animated.Value(0);
    this.state = {
      isKeyboardVisible: false,
    };
  }

  UNSAFE_componentWillMount() {
    this.keyboardWillShowSub = Keyboard.addListener(KeyboardEvent.Show, this.keyboardWillShow);
    this.keyboardWillHideSub = Keyboard.addListener(KeyboardEvent.Hide, this.keyboardWillHide);
  }

  componentWillUnmount() {
    this.keyboardWillShowSub.remove();
    this.keyboardWillHideSub.remove();
  }

  keyboardWillShow = (event) => {
    this.setState({ isKeyboardVisible: true }, () => {
      if (Platform.isAndroid) {
        return;
      }

      Animated.parallel([
        Animated.timing(this.keyboardHeight, {
          duration: event.duration,
          toValue: event.endCoordinates.height,
          useNativeDriver: false,
        }),
      ]).start();
    });
  };

  keyboardWillHide = (event) => {
    if (Platform.isAndroid) {
      this.setState({ isKeyboardVisible: false });
      return;
    }
    Animated.parallel([
      Animated.timing(this.keyboardHeight, {
        duration: event.duration,
        toValue: 0,
        useNativeDriver: false,
      }),
    ]).start(() => this.setState({ isKeyboardVisible: false }));
  };

  render() {
    const { children, style, ...props } = this.props;
    const { isKeyboardVisible } = this.state;
    const ptBottom = Platform.isAndroid ? 0 : this.keyboardHeight;
    return (
      <Animated.View style={[styles.container, { paddingBottom: ptBottom }, style]} {...props}>
        {typeof children === 'function' ? children(isKeyboardVisible) : children}
      </Animated.View>
    );
  }
}
