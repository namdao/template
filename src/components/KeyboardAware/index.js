import KeyboardAware from './KeyboardAware';

export default KeyboardAware;
