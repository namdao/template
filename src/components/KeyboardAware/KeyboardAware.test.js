import React from 'react';
import { Keyboard } from 'react-native';
import { shallow } from 'enzyme';
import KeyboardAware from './KeyboardAware';

const setup = (propOverrides) => {
  const props = { ...propOverrides };
  const wrapper = shallow(<KeyboardAware {...props} />);
  return {
    props,
    wrapper,
  };
};

Keyboard.addListener = jest.fn();

describe('KeyboardAware check render', () => {
  test('check componentWillMount', () => {
    const { wrapper } = setup();
    wrapper.instance().UNSAFE_componentWillMount();
    expect(Keyboard.addListener).toBeCalledTimes(4);
  });
  test('check unMount', () => {
    const { wrapper } = setup();
    wrapper.instance().UNSAFE_componentWillMount();
    wrapper.instance().keyboardWillShowSub = {
      remove: jest.fn(),
    };
    wrapper.instance().keyboardWillHideSub = {
      remove: jest.fn(),
    };
    wrapper.instance().componentWillUnmount();
    expect(wrapper.instance().keyboardWillShowSub.remove).toBeCalled();
    expect(wrapper.instance().keyboardWillHideSub.remove).toBeCalled();
  });
});

describe('KeyboardAware check snapShot', () => {
  test('should match snapshot', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
