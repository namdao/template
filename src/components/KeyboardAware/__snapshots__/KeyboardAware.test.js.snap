// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyboardAware check snapShot should match snapshot 1`] = `
<AnimatedComponent
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
      Object {},
    ]
  }
/>
`;
